api/src/auth_routes.ts

```diff
@@ -63,9 +64,35 @@ export function determine_callback_urls(provider_name: string): {
   };
 }
 
+/**
+ * Check that a redirect URL is one that we allow
+ *
+ * @param redirect URL to redirect to
+ * @returns a valid URl to redirect to, default to '/' if
+ *   the one passed in is bad
+ */
+function validateRedirect(redirect: string) {
+  if (redirect.startsWith('http')) {
+    // should match against a whitelist of allowed URLs
+    return redirect;
+  } else if (
+    redirect.startsWith('/') ||
+    redirect.startsWith('org.fedarch.faims3://')
+  ) {
+    return redirect;
+  } else {
+    return '/';
+  }
+}
+
 export function add_auth_routes(app: any, handlers: any) {
-  app.get('/auth/', (req: any, res: any) => {
-    // Allow the user to decide what auth mechanism to use
+  app.get('/auth/', async (req: any, res: any) => {
+    const redirect = validateRedirect(req.query?.redirect || '/');
+
+    if (req.user) {
+      await redirect_with_token(res, req.user, redirect);
+    }
+
     const available_provider_info = [];
```

Comment:
some inline comments in this function would be nice - pretty hard to understand what is going on